### ISSUE TITLE
Make SAW core parser handle pretty-printed output

### ISSUE THREAD

**Issue by eddywestbrook** [GaloisInc/saw-script#1595]
Currently, the SAW core pretty-printer will print out versions of SAW core terms that cannot be parsed back in by the SAW core parser. The main change required would be having the parser support the pretty-printers let-binding syntax which it uses to avoid exponential blowup, which looks like this:

```
let { x@1 = ...
      x@2 = ... } in e
```

Additionally, this would require extending the SAW core lexer to handle much more liberal names, including the `x@N` variable names and also the URI-based names generated for Cryptol identifiers and generated `ExtCns` symbols, since these are all currently printed by the pretty-printer but not parsed by the parser.

**Comment by danmatichuk** [comment]
As per the discussion with @brianhuffman below, I'll try implementing this as a simple syntactic de-sugaring. The main change to existing functionality will be to have SAW add type annotations to variables when it produces let-bindings during pretty-printing. This can then be parsed back in as an applied lambda abstraction (i.e `let { x : T = y } in e` -> `(\(x : T) -> e) y`.

Transcript:

Daniel Matichuk:

> My current thinking is that we'll need an additional Let-binding constructor for the AST, unless we want to require that all let-bindings have type annotations (which would allow for let-binding to just desugar into lambda abstraction). 
> 
> For type-checked terms we have a few options:
> 1) erase let-bindings during type checking (i.e zeta reduction)
> 2) leave let-bindings as un-reduced lambda applications
> 3) add an additional Let constructor to TermF
> 
> The disadvantage with 1) is that the pretty printer won't necessarily recover the same term shape, and won't have the provided variable names regardless. With 2) we have a chance to recover the names, until any beta-reduction occurs. With 3) we have the obvious complication of just needing to plumb an additional constructor through.
> 
> What do you think the best approach would be?

Brian Huffman:

> During pretty-printing, let bindings are generated automatically whenever the pretty printer notices repeated subterms in the same variable scope; they don't correspond to any actual AST nodes. The motivation for supporting let syntax is so that we can re-parse pretty-printed SAWCore (see [#1595](https://github.com/GaloisInc/saw-script/issues/1595)). This means we want option 1
> 
> That said, it would be nice to have some way to preserve the local variable names for let bindings. You could imagine having a global table or a special field on each Term that has an optional display name that the pretty printer could use instead of using a machine-generated name like x@1

Daniel Matichuk:

> Okay, so it sounds like we can address those somewhat independently, where let-bindings are just used to populate some name hint field but are otherwise not relevant to the actual term representation.

> As for the parser itself, from what I can tell the pretty printer doesn't add type annotations to let-bindings (currently), meaning that we can't simply desugar them into lambdas.

Brian Huffman:
> As of recently, every SAWCore term is annotated internally with its type, so there's no technical obstacle to doing that

Daniel Matichuk:
> Alright, so it sounds like the easiest approach is to just modify the pretty printer to add type annotations, which then allows the parser to just de-sugar let-bindings into lambdas. That basically allows this to be a purely syntactic change.

**Comment by brianhuffman** [comment]
The import property we want to satisfy is that for any term `t`, `parse(print(t)) = t`.

The pretty printer invents let bindings wherever there are shared subterms, so e.g. a term like `f (g a) (g a)` might be printed out as `let x@1 = g a in f x@1 x@1`. If we parsed let bindings as lambdas, the parser would re-parse this as `(\x -> f x x) (g a)`, which is not what we want. To correctly reconstruct the original term, let bindings need to be inlined.

**Comment by danmatichuk** [comment]
Yes I agree, which I think requires at least adding Let as a constructor to SAWCore.Parser.UTerm, so that it can be properly expanded during type inference/checking. Basically it ends up being a Lambda with an implicit rule to always beta-reduce.

**Comment by sauclovian-g** [comment]
Adding things to the untyped AST has very little downstream cost :-)

**Comment by danmatichuk** [comment]
@brianhuffman Do you have examples for the additional identifier forms that need to be supported in addition to `x@N`?

**Comment by sauclovian-g** [comment]
I don't think this is actually done, that was just the major part of it

**Comment by sauclovian-g** [comment]
Anyway, there's at least the potential for a full URI in there. We keep tossing around the idea of changing that a bit to suit our needs better (the URI library we're using was convenient at one point, but it's not really a good match to the semantics we want) and we should probably sort out the scheme we _do_ want before adding concrete syntax for it.

**Comment by brianhuffman** [comment]
I'd be happy to rename this issue to **Make SAW core parser handle pretty-printed let expressions** and close it. We have some other open issues about parsing SAWCore names, e.g. #2668 and #330.

**Comment by sauclovian-g** [comment]
Right, that's fine, but we do want the actual round trip to work and I'm not sure we have a handle yet on what else is missing. Maybe we should add some roundtrip tests that start with some more complex SAWCore terms (e.g. things derived from Cryptol), make new issues for any problems those expose, and then close this one...

**Comment by sauclovian-g** [comment]
We could for example rewrite (or copy) test0057 to use the concrete syntax instead of the extcore format as a starting point.

That however leads to some other issues with being able to replace the extcore format, like saving or reconstructing Cryptol types for terms.
